### ISSUE TITLE
deeplink destroys the default route

### ISSUE THREAD

**Issue by mofengfs** [expo/expo#12067]
### Summary

Expo bare uses reactnavigation v5's deeplink configuration. The default route is destroyed.

When I open the app manually（Important: I am not using a deep link to open the app. Many people mistakenly think that I am using deeplink to open the app.）, no matter which page I go to, the Android backbutton is to exit the app instead of back to the previous page.

Since the navigation team can't give an answer, I suspect it's Expo bare's problem.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

40

### Environment

Expo CLI 4.1.6 environment info:
    System:
      OS: Windows 10 10.0.18363
    Binaries:
      Node: 14.15.1 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.10 - C:\Users\Administrator\AppData\Roaming\npm\yarn.CMD
      npm: 6.14.11 - C:\Program Files\nodejs\npm.CMD
    SDKs:
      Android SDK:
        API Levels: 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30
        Build Tools: 23.0.3, 24.0.3, 25.0.3, 26.0.3, 28.0.3, 29.0.2, 29.0.3, 30.0.3, 31.0.0
    IDEs:
      Android Studio: Version  4.1.0.0 AI-201.8743.12.41.6953283
    npmPackages:
      expo: ~40.0.0 => 40.0.0
      react: 16.13.1 => 16.13.1
      react-dom: 16.13.1 => 16.13.1
      react-native: 0.63.4 => 0.63.4
      react-native-web: ~0.13.12 => 0.13.18
    Expo Workflow: bare

> react navigation info
"@react-navigation/native": "^5.9.2",
"@react-navigation/stack": "^5.14.2",

### Reproducible demo or steps to reproduce from a blank project

> android/app/src/main/src/AndroidManifest.xml

![image](https://user-images.githubusercontent.com/16655229/109443833-92a5c880-7a76-11eb-8057-49410c6596a3.png)

> deeplink config

![image](https://user-images.githubusercontent.com/16655229/109443843-9e918a80-7a76-11eb-913a-e0bfefafb0ab.png)

> route

![image](https://user-images.githubusercontent.com/16655229/109443860-b0732d80-7a76-11eb-9773-dd429e7d0a95.png)

> test

`npx uri-scheme open a08://warn/1 --android`

![image](https://user-images.githubusercontent.com/16655229/109443900-caad0b80-7a76-11eb-8149-469ba5d24c48.png)

> Open the app manually (important) to enter other pages. Press backbutton

![image](https://user-images.githubusercontent.com/16655229/109447981-ed8fed80-7a7f-11eb-930a-ca62d84ff747.png)

**Comment by mofengfs** [comment]
Supplement: Even if I set initialRouteName, it doesn't work.

**Comment by brentvatne** [comment]
you should post this to the react-navigation repository and share a minimal reproducible example (https://stackoverflow.com/help/minimal-reproducible-example). an expo bare project is just a plain react-native project with expo modules added on, and nothing expo-specific about it (and nothing in this repository) would impact the deep linking behavior.

**Comment by brentvatne** [comment]
here's a project that you can clone and run and verify that as you switch tabs and press the back button it works as expected: https://github.com/brentvatne/mofengfs

**Comment by mofengfs** [comment]
@brentvatne 

Thank you, but
1. I used `npx uri-scheme open a08://two` and failed to open the 'two' screen.
2. I create a TestScreen in the project, and perform stack jump in ‘one’.
I use `npx uri-scheme open a08://test`, but the test interface cannot be opened.

> deeplink

```javascript
import * as Linking from 'expo-linking';

export default {
  prefixes: [Linking.makeUrl('/')],
  config: {
    screens: {
      Root: {
        screens: {
          TabOne: {
            screens: {
              TabOneScreen: 'one',
              Test: {
                screens: {
                  TestScreen: 'test'
                }
              }
            },
          },
          TabTwo: {
            screens: {
              TabTwoScreen: 'two',
            },
          },
        },
      },
      NotFound: '*',
      
    },
  },
};

```
or


```javascript
import * as Linking from 'expo-linking';

export default {
  prefixes: [Linking.makeUrl('/')],
  config: {
    screens: {
      Root: {
        screens: {
          TabOne: {
            screens: {
              TabOneScreen: 'one',
            },
          },
          TabTwo: {
            screens: {
              TabTwoScreen: 'two',
            },
          },
        },
      },
      NotFound: '*',
      Test: {
        screens: {
          TestScreen: 'test'
        }
      }
    },
  },
};


```

> TabOneScreen

![image](https://user-images.githubusercontent.com/16655229/109600471-24d1cd80-7b58-11eb-8d9e-9387e1bcda24.png)

> TestScreen

![image](https://user-images.githubusercontent.com/16655229/109600603-5054b800-7b58-11eb-98c7-235959e66203.png)

**Comment by brentvatne** [comment]
@mofengfs - change https://github.com/brentvatne/mofengfs/blob/master/navigation/LinkingConfiguration.ts#L4 to `prefixes: ['a08://'],`

you will notice that this works fine, and you can reproduce the back button behavior you describe, which is nothing to do with any of this code in this repo

**Comment by mofengfs** [comment]
Hi @brentvatne , many thx

In order to restore my project to a high degree, I fork the project you created, my fork: https://github.com/mofengfs/mofengfs-1.git, modify and test it.
I excluded `Expo eject`, `UI-kitten ApplicationProvider`, `deeplink config`...
In the end, I discovered that `BackHandler` caused the default route to be destroyed after deeplink was configured.
But I still don't know why this is.


![image](https://user-images.githubusercontent.com/16655229/109926855-68663c00-7cfe-11eb-966d-4a2a77c95f71.png)
